src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -349,5 +411,152 @@ internal static bool TryParseSeverity(string value, out ReportDiagnostic severit
             severity = default;
             return false;
         }
+
+        internal static bool IsPartialGlobalConfig(AnalyzerConfig config) => config.GlobalSection.Properties.ContainsKey(GlobalKey);
+
+        /// <summary>
+        /// Merge any partial global configs into a single global config, and remove the partial configs
+        /// </summary>
+        /// <param name="analyzerConfigs">An <see cref="ArrayBuilder{T}"/> of <see cref="AnalyzerConfig"/> containing a mix of regular and unmerged partial global configs</param>
+        /// <param name="diagnostics">Diagnostics produced during merge will be added to this bag</param>
+        /// <returns>A <see cref="GlobalAnalyzerConfig" /> that contains the merged partial configs, or <c>null</c> if there were no partial configs</returns>
+        internal static GlobalAnalyzerConfig? MergeGlobalConfigs(ArrayBuilder<AnalyzerConfig> analyzerConfigs, DiagnosticBag diagnostics)
+        {
+            GlobalAnalyzerConfigBuilder globalAnalyzerConfigBuilder = new GlobalAnalyzerConfigBuilder();
+            for (int i = 0; i < analyzerConfigs.Count; i++)
+            {
+                if (IsPartialGlobalConfig(analyzerConfigs[i]))
+                {
+                    globalAnalyzerConfigBuilder.MergeIntoGlobalConfig(analyzerConfigs[i]);
+                    analyzerConfigs.RemoveAt(i);
+                    i--;
+                }
+            }
+
+            var globalConfig = globalAnalyzerConfigBuilder.Build(diagnostics);
+            return globalConfig;
+        }
+
+        /// <summary>
+        /// Builds a global analyzer config from a series of partial configs
+        /// </summary>
+        class GlobalAnalyzerConfigBuilder
+        {
+            private PooledDictionary<string, ImmutableDictionary<string, (string value, string configPath)>.Builder>? _values;
+            private PooledDictionary<string, ImmutableDictionary<string, ArrayBuilder<string>>.Builder>? _duplicates;
+
+            internal void MergeIntoGlobalConfig(AnalyzerConfig config)
+            {
+                if (_values is null)
+                {
+                    _values = PooledDictionary<string, ImmutableDictionary<string, (string, string)>.Builder>.GetInstance();
+                    _duplicates = PooledDictionary<string, ImmutableDictionary<string, ArrayBuilder<string>>.Builder>.GetInstance();
+                }
+
+                MergeSection(config.PathToFile, config.GlobalSection);
+                foreach (var section in config.NamedSections)
+                {
+                    MergeSection(config.PathToFile, section);
+                }
+            }
+
+            internal GlobalAnalyzerConfig? Build(DiagnosticBag diagnostics)
```

Comment:
Did you consider guarding against multiple calls to `Build` here? Basically an assert or something? Or is it just not worth it?

Reply:
Yeah, I wasn't sure it was worth it. But I've changed it to null out the `_values` and `_duplicates` calls after building, so subsequent calls will just return `null` rather than throwing.

---
In reply to: [423049256](https://github.com/dotnet/roslyn/pull/43889#discussion_r423049256) [](ancestors = 423049256)